### PR TITLE
Pod ENI

### PR DESCRIPTION
**Pull request by bdwyertech**
**1. Issue, if available:**
#905 

**2. Description of changes:**
This improves instance type selection to factor in [Security Groups for Pods](https://docs.aws.amazon.com/eks/latest/userguide/security-groups-for-pods.html) by ensuring only instance types capable of Pod ENI are launched if set in a pods resource request.

This leverages the lookup table in [vpc-resource-controller](https://github.com/aws/amazon-vpc-resource-controller-k8s/blob/05e89ff9300a5cc0ebea705834cf27f0a7f3b509/pkg/aws/vpc/limits.go#L27) to determine Pod ENI support.

**3. Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: *link to issue*
- [ ] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by ellistarn** [comment]
This is a pretty cool WIP! I'm assuming that `aws.amazon.com/pod-eni` is already written onto nodes by the CNI controller and works w/ the kube scheduler?

**Comment by bdwyertech** [comment]
> This is a pretty cool WIP! I'm assuming that `aws.amazon.com/pod-eni` is already written onto nodes by the CNI controller and works w/ the kube scheduler?

So the namespace is up for debate.  When the CNI first initializes, it sets `vpc.amazonaws.com/has-trunk-attached=false` and stays this way until a pod runs on it which requires a branch interface, then it flips over to `true`.  This label cannot really be used to make a determination of Pod ENI support.

If we want to stay out of the `aws.amazon.com` namespace, maybe `aws.karpenter.sh/pod-eni` or `aws.karpenter.sh/pod-eni-capable` would be more appropriate.

I was leveraging the initial node template to set the value initially and then reverted it back.

https://github.com/aws/karpenter/blob/e8537a3622dac212de4bb7a68f31c07dbcdbe8fd/pkg/cloudprovider/aws/instance.go#L240-L250

**Comment by ellistarn** [comment]
Awesome! This is shaping up really nicely!

**Comment by bdwyertech** [comment]
I was kind of surprised by the hard-coded lookup table in the CNI and that the new m6i and m6a instances don't support this!

**Comment by ellistarn** [comment]
Ready for review?

File: pkg/cloudprovider/aws/suite_test.go [hunk 1]
```diff
@@ -121,6 +121,20 @@ var _ = Describe("Allocation", func() {
 
 	Context("Reconciliation", func() {
 		Context("Specialized Hardware", func() {
+			It("should not launch AWS Pod ENI on a t3", func() {
```
Comment (ellistarn) [inline pkg/cloudprovider/aws/suite_test.go:124]: Can you have a positive test that shows that it does schedule w/ this requirement?

**Review by ellistarn — COMMENTED** [review]
One minor comment. Otherwise great job!

File: pkg/controllers/provisioning/binpacking/packable.go [hunk 2]
```diff
@@ -242,7 +242,7 @@ func (p *Packable) validateAWSNeurons(pods []*v1.Pod) error {
 func (p *Packable) validateAWSPodENI(pods []*v1.Pod) error {
 	for _, pod := range pods {
 		for _, container := range pod.Spec.Containers {
-			if _, ok := container.Resources.Requests[resources.AWSPodENI]; ok {
+			if _, ok := container.Resources.Limits[resources.AWSPodENI]; ok {
```
Comment (ellistarn) [inline pkg/controllers/provisioning/binpacking/packable.go:245]: I think you want to use Requests here. Kubernetes will default requests to the value of limits if only limits are set.

File: pkg/cloudprovider/aws/suite_test.go (hunk 1 above)
Comment (bdwyertech) [reply pkg/cloudprovider/aws/suite_test.go:125]: Done!

File: pkg/controllers/provisioning/binpacking/packable.go (hunk 2 above)
Comment (bdwyertech) [reply pkg/controllers/provisioning/binpacking/packable.go]: Yup, I tested and you're right.  swapped it back.  Saw an issue when i set requests but not limit on the client side.

**Review by ellistarn — APPROVED** [review]
Nicely done!

**Pull request merged**